ServiceTypesQuery.test_integration_with_listener is flakey

Likely fixed by https://github.com/jstasiak/python-zeroconf/pull/419
```
_______________ ServiceTypesQuery.test_integration_with_listener _______________

self = <zeroconf.test.ServiceTypesQuery testMethod=test_integration_with_listener>

    def test_integration_with_listener(self):
    
        type_ = "_test-srvc-type._tcp.local."
        name = "xxxyyy"
        registration_name = "%s.%s" % (name, type_)
    
        zeroconf_registrar = Zeroconf(interfaces=['127.0.0.1'])
        desc = {'path': '/~paulsm/'}
        info = ServiceInfo(
            type_,
            registration_name,
            80,
            0,
            0,
            desc,
            "ash-2.local.",
            addresses=[socket.inet_aton("10.0.1.2")],
        )
        zeroconf_registrar.register_service(info)
    
        try:
            service_types = ZeroconfServiceTypes.find(interfaces=['127.0.0.1'], timeout=0.5)
>           assert type_ in service_types
E           AssertionError: assert '_test-srvc-type._tcp.local.' in ('_net-assistant._udp.local.', '_rfb._tcp.local.', '_sftp-ssh._tcp.local.', '_ssh._tcp.local.')

zeroconf/test.py:1159: AssertionError
```

--- a/zeroconf/__init__.py
+++ b/zeroconf/__init__.py
@@ -1785,17 +1785,33 @@ class ServiceBrowser(_ServiceBrowserBase, threading.Thread):
         super().cancel()
         self.join()
 
+    def _wait_for_next_event(self) -> None:
+        """Wait for the next handler or time to send queries."""
+        # If there are handlers to call
+        # we want to process them right away
+        if self._handlers_to_call:
+            return
+
+        # Wait for the type has the smallest next time
+        next_time = min(self._next_time.values())
+        now = current_time_millis()
+
+        if next_time <= now:
+            return
+
+        with self.zc.condition:
+            # We must check again while holding the condition
+            # in case the other thread has added to _handlers_to_call
+            # between when we checked above when we were not
+            # holding the condition
+            if not self._handlers_to_call:
+                self.zc.condition.wait(millis_to_seconds(next_time - now))
+
     def run(self) -> None:
         """Run the browser thread."""
         super().run()
         while True:
-            if not self._handlers_to_call:
-                # Wait for the type has the smallest next time
-                next_time = min(self._next_time.values())
-                now = current_time_millis()
-                if next_time > now:
-                    self.zc.wait(next_time - now)
-
+            self._wait_for_next_event()
             if self.zc.done or self.done:
                 return
 

